--[[
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::::::::'##:::'##::'#######::'##::::'##:'##:::::::'####::'######:::::::::
:::::::: ##::'##::'##.... ##: ##:::: ##: ##::::::: ####:'##... ##::::::::
:::::::: ##:'##::: ##:::: ##: ##:::: ##: ##:::::::. ##:: ##:::..:::::::::
:::::::: #####:::: ##:::: ##: #########: ##:::::::'##:::. ######:::::::::
:::::::: ##. ##::: ##:::: ##: ##.... ##: ##:::::::..:::::..... ##::::::::
:::::::: ##:. ##:: ##:::: ##: ##:::: ##: ##:::::::::::::'##::: ##::::::::
:::::::: ##::. ##:. #######:: ##:::: ##: ########:::::::. ######:::::::::
::::::::..::::..:::.......:::..:::::..::........:::::::::......::::::::::
:::::::::::::::'###::::'########::'##::::'##:'####:'##::: ##:::::::::::::
::::::::::::::'## ##::: ##.... ##: ###::'###:. ##:: ###:: ##:::::::::::::
:::::::::::::'##:. ##:: ##:::: ##: ####'####:: ##:: ####: ##:::::::::::::
::::::::::::'##:::. ##: ##:::: ##: ## ### ##:: ##:: ## ## ##:::::::::::::
:::::::::::: #########: ##:::: ##: ##. #: ##:: ##:: ##. ####:::::::::::::
:::::::::::: ##.... ##: ##:::: ##: ##:.:: ##:: ##:: ##:. ###:::::::::::::
:::::::::::: ##:::: ##: ########:: ##:::: ##:'####: ##::. ##:::::::::::::
::::::::::::..:::::..::........:::..:::::..::....::..::::..::::::::::::::
:::'####:'##::: ##:'########:'####:'##::: ##:'####:'########:'########:::
:::. ##:: ###:: ##: ##.....::. ##:: ###:: ##:. ##::... ##..:: ##.....::::
:::: ##:: ####: ##: ##:::::::: ##:: ####: ##:: ##::::: ##:::: ##:::::::::
:::: ##:: ## ## ##: ######:::: ##:: ## ## ##:: ##::::: ##:::: ######:::::
:::: ##:: ##. ####: ##...::::: ##:: ##. ####:: ##::::: ##:::: ##...::::::
:::: ##:: ##:. ###: ##:::::::: ##:: ##:. ###:: ##::::: ##:::: ##:::::::::
:::'####: ##::. ##: ##:::::::'####: ##::. ##:'####:::: ##:::: ########:::
:::....::..::::..::..::::::::....::..::::..::....:::::..:::::........::::
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
:::::::::::: Thanks for using Kohl's Admin Infinite Edition! ::::::::::::
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
:::::::::::::::::::: This script is developed by Kohl :::::::::::::::::::
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: To configure the admin open the Settings module inside of this script	:
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
]]

if _G.KAU then
	script:Destroy()
else
	_G.KAU = true
	shared._K_LOADER_LOADED = true
	script.Name = "Kohl's Admin Infinite"
	script.Parent = game:GetService("ServerScriptService")

	local function loadModuleAsset(assetId: number)
		local asset = game:GetService("InsertService"):LoadAsset(assetId)
		local module = asset:FindFirstChild("MainModuleKAI", true)
		if module then
			return assert(require(module), ":LoadAsset() module returned invalid values!")
		end
		error(`Failed to find the source for {assetId}.`)
	end

	local loaderId = 172732271
	local moduleId = 1868400649
	local ok, result = pcall(require, moduleId)
	if not ok then
		warn("Kohl's Admin failed to auto-update from MainModule, attempting to load from asset\nReason:\t" .. result)
		ok, result = pcall(loadModuleAsset, loaderId)
		if not ok then
			warn("Kohl's Admin failed to auto-update from asset, attempting to load from local package\nReason:\t" .. result)
			result = require(script:WaitForChild("MainModuleKAI"))
		end
	end
end
